openpype/plugins/publish/integrate_new.py

```diff
@@ -110,801 +156,494 @@ class IntegrateAssetNew(pyblish.api.InstancePlugin):
+            assert (
+                len(src_collection.indexes) == len(dst_collection.indexes)
+            ), "This is a bug"
+
+            # Multiple file transfers
+            transfers = []
+            for src_file_name, dst in zip(src_collection, dst_collection):
+                src = os.path.join(stagingdir, src_file_name)
+                transfers.append((src, dst))
 
-        try:
-            filled_template = StringTemplate.format_strict_template(
-                template, fill_pairs
+        else:
+            # Single file
+            fname = files
+            assert not os.path.isabs(fname), (
+                "Given file name is a full path"
             )
-        except (KeyError, TemplateUnsolved):
-            keys = []
-            if fill_pairs:
-                keys = fill_pairs.keys()
 
-            msg = "Subset grouping failed. " \
-                  "Only {} are expected in Settings".format(','.join(keys))
-            self.log.warning(msg)
+            # Manage anatomy template data
+            template_data.pop("frame", None)
+            if repre.get("udim"):
+                template_data["udim"] = repre["udim"][0]
 
-        return filled_template
+            # Construct destination filepath from template
+            anatomy_filled = anatomy.format(template_data)
+            template_filled = anatomy_filled[template_name]["path"]
+            repre_context = template_filled.used_values
+            dst = os.path.normpath(template_filled)
 
-    def create_version(self, subset, version_number, data=None):
-        """ Copy given source to destination
+            # Single file transfer
+            src = os.path.join(stagingdir, fname)
+            transfers = [(src, dst)]
 
-        Args:
-            subset (dict): the registered subset of the asset
-            version_number (int): the version number
+        for key in self.db_representation_context_keys:
+            value = template_data.get(key)
+            if not value:
+                continue
+            repre_context[key] = template_data[key]
 
-        Returns:
-            dict: collection of data to create a version
-        """
+        # Explicitly store the full list even though template data might
+        # have a different value
+        if repre.get("udim"):
+            repre_context["udim"] = repre.get("udim")  # store list
 
-        return {"schema": "openpype:version-3.0",
-                "type": "version",
-                "parent": subset["_id"],
-                "name": version_number,
-                "data": data}
+        # Use previous representation's id if there is a name match
+        existing = existing_repres_by_name.get(repre["name"].lower())
+        if existing:
+            repre_id = existing["_id"]
+        else:
+            repre_id = ObjectId()
+
+        # Backwards compatibility:
+        # Store first transferred destination as published path data
+        # todo: can we remove this?
```

Comment:
Following integrators need to know the final destination of each representation without having to re-calculate it. So unless we find a better way, I don't think we can remove it.

Reply:
> Following integrators need to know the final destination of each representation without having to re-calculate it. So unless we find a better way, I don't think we can remove it.

We're already storing the 'transfers' or `files` data in the actual representation entities. We might be better of using either of those data then duplicating the data and also storing `published_files` separately?

Reply:
We have to know which files were published for the representation not for instance.

Reply:
> We have to know which files were published for the representation not for instance.

That would still be the case, because those `files` list would also only have the files related to that particular representation. Albeit they would ALSO contain the `instance.data["transfers"]` since those files are also relevant to that representation. As shown in [this example](https://github.com/pypeclub/OpenPype/pull/2898#discussion_r840022014).

(Note that currently it does note store `instance.data["transfers"]` as files after the refactor at all and thus Site Sync would not work for those.)

Reply:
`published_path` is not aboutthe file itself, but rather being able to tell later on in the process where this publish in general ended up being integrated into. There is no other place where to get that information. And it needs to be the "main" file of the repre. If we're looking just at the `representation['files']` list, that you can't be sure which one of those files is representative of the publish location really, considering some might be a folder deeper in `resources` subfolder. 

This is another piece that could be solved by more robust `files` data on the repre though. As mentioned here https://github.com/pypeclub/OpenPype/pull/2898/files/a7a908d1348381ab0c4df9c29861d7c02be635cb#r839325287

To give a simple pseudoexample:

```
representation["files"] = [
	{
	source: "my_original_filename.abc", 							# filled before the integrator
	transfer_type: "copy", 										# filled before the integrator
	destination: "path/to/final/integrated/directory/new_filename.abc", 	# filled by integrator
	resource_type: "primary", 									# filled before the integrator
	},
	{
	source: "additional_texture.exr",
	transfer_type: "hardlink",
	destination: "path/to/final/integrated/directory/resources/new_filename.exr",
	resource_type: "auxiliary",
	}
]
```

That could cover all use cases of `transfers`, `hardlinks`, `published_path` and would also possibly match the files data model in the DB exactly and could be inserted (with exter keys like `created_dt`) directly into the db. It would also allow distinguishing between the `main` representation files and `resources` even from the db record, rather than just a simple list of files.

Reply:
What purpose have `transfer_type` and `resource_type`? 
I guess "gimme all hardlinks", "gimme all primary files", is there anything else we would need for it to be useful? (Just brainstorming here.)

Reply:
Well that `transfer_type` would be needed for the publishing process itself so the integrator know whether to copy or hardlink. Not really important for the DB, however, useful to query how much stuff we copy vs hardlink when looking into the project data in general. 

`resource_type` could be used by loaders to determine how to approach the various files. For instance look loader now simply expects that resources are in the `resources` subfolder in the publish (hardcoded), but with this info the loader could actually know what is the main shader file, vs what are the textures needed for it.